README.md
This library provide simple encryption and decryption


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

---

## 1.0.0 - 31 Mar 2020

### Added

- First release

---


examples/example.php
<?php
// Example code
include_once('./../../../autoload.php');

/**
 * Pretty print array/object for debuging
 *
 * @param array|object $params Array/object to be print
 * @param boolean $exit Exit after print
 * @return void
 */
if (!function_exists('\pr')) {
    function pr($params, $exit = true)
    {
        echo "<pre>";
        print_r($params);
        echo "</pre>";

        if ($exit == true) {
            exit();
        }
    }
}

// Create $cryptor object
$config =  array(
    'encrypt_method' => 'AES-256-CBC',
    'iterations' => 999
);
$cryptor = new \FR\Cryptor\Cryptor($config);
?>

<!DOCTYPE html>
<html>

<body>
    <?php
    // How to check instance type
    if ($cryptor instanceof \FR\Cryptor\CryptorInterface)
        echo 'How to check instance type: \FR\Cryptor\CryptorInterface <br>';

    // How to encrypt
    $string = 'Encrypt this string from PHP';
    $key = '123';
    $encrypted_string = $cryptor->encrypt($string, $key);
    echo '<b>Encrypted string from PHP:</b> ' . $encrypted_string;
    echo '<br>';

    // How to decrypt
    $orignal_string = $cryptor->decrypt($encrypted_string, $key);
    echo '<b>Orignal string decrypt from PHP:</b> ' . $orignal_string;
    echo '<br>';
    ?>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/3.1.9-1/crypto-js.js"></script>
    <script src="./js/fr-cryptor.js"></script>
    <script>
        // TODO: fr-cryptor should work on config itrations and encrypt_methods
        var string = 'Encrypt this string from JS';
        var key = '456';

        let cryptor = new FRCryptor();
        var encrypted_string = cryptor.encrypt(string, key);
        console.log('Encrypted string from JS');
        console.log(encrypted_string);

        var orignal_string = cryptor.decrypt(encrypted_string, key);
        console.log('Orignal string decrypt from JS');
        console.log(orignal_string);

        var orignal_string = cryptor.decrypt('<?php echo $encrypted_string; ?>', '<?php echo $key; ?>');
        console.log('Orignal string of PHP decrypted from JS');
        console.log(orignal_string);
    </script>
</body>

</html>

examples/js/fr-cryptor.js
class FRCryptor {
  /**
   * @var integer Return encrypt method or Cipher method number. (128, 192, 256)
   */
  get encryptMethodLength() {
    var encryptMethod = this.encryptMethod;
    // get only number from string.
    // @link https://stackoverflow.com/a/10003709/128761 Reference.
    var aesNumber = encryptMethod.match(/\d+/)[0];
    return parseInt(aesNumber);
  } // encryptMethodLength

  /**
   * @var integer Return cipher method divide by 8. example: AES number 256 will be 256/8 = 32.
   */
  get encryptKeySize() {
    var aesNumber = this.encryptMethodLength;
    return parseInt(aesNumber / 8);
  } // encryptKeySize

  /**
   * @link http://php.net/manual/en/function.openssl-get-cipher-methods.php Refer to available methods in PHP if we are working between JS & PHP encryption.
   * @var string Cipher method.
   *              Recommended AES-128-CBC, AES-192-CBC, AES-256-CBC
   *              due to there is no `openssl_cipher_iv_length()` function in JavaScript
   *              and all of these methods are known as 16 in iv_length.
   */
  get encryptMethod() {
    return 'AES-256-CBC';
  } // encryptMethod

  /**
   * Decrypt string.
   * @param string encryptedString The encrypted string to be decrypt.
   * @param string key The key.
   * @return string Return decrypted string.
   */
  decrypt(encryptedString, key) {
    var json = JSON.parse(
      CryptoJS.enc.Utf8.stringify(CryptoJS.enc.Base64.parse(encryptedString))
    );

    var salt = CryptoJS.enc.Hex.parse(json.salt);
    var iv = CryptoJS.enc.Hex.parse(json.iv);

    var encrypted = json.ciphertext; // no need to base64 decode.

    var iterations = parseInt(json.iterations);
    if (iterations <= 0) {
      iterations = 999;
    }
    var encryptMethodLength = this.encryptMethodLength / 4; // example: AES number is 256 / 4 = 64
    var hashKey = CryptoJS.PBKDF2(key, salt, {
      hasher: CryptoJS.algo.SHA512,
      keySize: encryptMethodLength / 8,
      iterations: iterations
    });

    var decrypted = CryptoJS.AES.decrypt(encrypted, hashKey, {
      mode: CryptoJS.mode.CBC,
      iv: iv
    });

    return decrypted.toString(CryptoJS.enc.Utf8);
  } // decrypt

  /**
   * Encrypt string.
   * @param string string The original string to be encrypt.
   * @param string key The key.
   * @return string Return encrypted string.
   */
  encrypt(string, key) {
    var iv = CryptoJS.lib.WordArray.random(16); // the reason to be 16, please read on `encryptMethod` property.

    var salt = CryptoJS.lib.WordArray.random(256);
    var iterations = 999;
    var encryptMethodLength = this.encryptMethodLength / 4; // example: AES number is 256 / 4 = 64
    var hashKey = CryptoJS.PBKDF2(key, salt, {
      hasher: CryptoJS.algo.SHA512,
      keySize: encryptMethodLength / 8,
      iterations: iterations
    });

    var encrypted = CryptoJS.AES.encrypt(string, hashKey, {
      mode: CryptoJS.mode.CBC,
      iv: iv
    });
    var encryptedString = CryptoJS.enc.Base64.stringify(encrypted.ciphertext);

    var output = {
      ciphertext: encryptedString,
      iv: CryptoJS.enc.Hex.stringify(iv),
      salt: CryptoJS.enc.Hex.stringify(salt),
      iterations: iterations
    };

    return CryptoJS.enc.Base64.stringify(CryptoJS.enc.Utf8.parse(JSON.stringify(output)));
  } // encrypt
}


src/Cryptor.php
<?php

namespace FR\Cryptor;

/**
 * @author Faisal Rehman <faisalrehmanid@hotmail.com>
 * 
 * This class provide abstraction layer of encryption and decryption 
 * and always create object of type \FR\Cryptor\CryptorInterface
 * 
 * Example: How to use this class?
 * 
 * ```
 * <?php
 *   $config =  array('encrypt_method' => 'AES-256-CBC',
 *                    'iterations' => 999);
 *   $cryptor = new \FR\Cryptor\Cryptor($config);
 *   $string = 'Sample string'; $key = 'sample-key';
 *   $encrypted_string = $cryptor->encrypt($string, $key);
 *   $orignal_string = $cryptor->encrypt($encrypted_string, $key);
 * ?>
 * ```
 */

class Cryptor implements CryptorInterface
{
    /**
     * Encrypt method
     *
     * Recommended methods: AES-128-CBC | AES-192-CBC | AES-256-CBC
     * @link http://php.net/manual/en/function.openssl-get-cipher-methods.php List of available encrypt methods
     * @var string
     */
    protected $encrypt_method;

    /**
     * Number of iterations to create complexity
     * More iterations means more processing
     *
     * @var int
     */
    protected $iterations;

    /**
     * Create object of \FR\Cryptor\CryptorInterface
     * 
     * // Cryptor configuration
     * $config =  array('encrypt_method' => 'AES-256-CBC',
     *                  'iterations' => 999);
     * Recommended methods: AES-128-CBC | AES-192-CBC | AES-256-CBC
     * 
     * @param array $config Cryptor configuration
     * @link http://php.net/manual/en/function.openssl-get-cipher-methods.php List of available encrypt methods
     * @throws \Exception `encrypt_method` is required in $config
     * @return object \FR\Cryptor\CryptorInterface
     */
    public function __construct(array $config)
    {
        @$encrypt_method = $config['encrypt_method'];
        @$iterations = $config['iterations'];

        if (!$encrypt_method)
            throw new \Exception('`encrypt_method` is required in $config');

        if (!$iterations || !is_int($iterations))
            throw new \Exception('`iterations` is required in $config must be integer');

        $this->encrypt_method = $encrypt_method;
        $this->iterations = $iterations;
    }

    /**
     * Get encrypt method length number: 128 | 192 | 256
     * @return int length
     */
    protected function encryptMethodLength()
    {
        $number = filter_var($this->encrypt_method, FILTER_SANITIZE_NUMBER_INT);

        return intval(abs($number));
    }

    /**
     * Apply encryption to the given string
     * @param string $string String to be encrypt
     * @param string $key Use to encrypt string
     * @return string Encrypted string
     */
    public function encrypt($string, $key)
    {
        $iv_length = openssl_cipher_iv_length($this->encrypt_method);
        $iv = openssl_random_pseudo_bytes($iv_length);

        $salt = openssl_random_pseudo_bytes(256);
        $hash_key = hash_pbkdf2('sha512', $key, $salt, $this->iterations, ($this->encryptMethodLength() / 4));

        $encrypted_string = openssl_encrypt(
            $string,
            $this->encrypt_method,
            hex2bin($hash_key),
            OPENSSL_RAW_DATA,
            $iv
        );

        $encrypted_string = base64_encode($encrypted_string);
        unset($hash_key);

        $output = [
            'ciphertext' => $encrypted_string,
            'iv'         => bin2hex($iv),
            'salt'       => bin2hex($salt),
            'iterations' => $this->iterations,
        ];
        unset($encrypted_string, $iv, $iv_length, $salt);

        return base64_encode(json_encode($output));
    }

    /**
     * Decrypt the given encrypted string
     * @param string $encrypted_string Encrypted string that is base64 encoded
     * @param string $key Use to decrypt string
     * @return mixed Original string
     */
    public function decrypt($encrypted_string, $key)
    {
        $json = json_decode(base64_decode($encrypted_string), true);

        try {
            $salt = hex2bin($json["salt"]);
            $iv = hex2bin($json["iv"]);
        } catch (\Exception $e) {
            return null;
        }

        $cipher_text = base64_decode($json['ciphertext']);

        $iterations = intval(abs($json['iterations']));
        $hash_key = hash_pbkdf2('sha512', $key, $salt, $iterations, ($this->encryptMethodLength() / 4));
        unset($iterations, $json, $salt);

        $decrypted = openssl_decrypt(
            $cipher_text,
            $this->encrypt_method,
            hex2bin($hash_key),
            OPENSSL_RAW_DATA,
            $iv
        );
        unset($cipher_text, $hash_key, $iv);

        return $decrypted;
    }
}
